modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseReaderLazyLoader.java

```diff
@@ -123,7 +133,49 @@ InputStream databaseInputStream() throws IOException {
         return Files.newInputStream(databasePath);
     }
 
-    DatabaseReader get() throws IOException {
+    CityResponse getCity(InetAddress ipAddress) {
+        SpecialPermission.check();
+        return AccessController.doPrivileged((PrivilegedAction<CityResponse>) () ->
+            cache.putIfAbsent(ipAddress, databasePath.toString(), ip -> {
+                try {
+                    return get().city(ip);
+                } catch (AddressNotFoundException e) {
+                    throw new GeoIpProcessor.AddressNotFoundRuntimeException(e);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }));
+    }
+
+    CountryResponse getCountry(InetAddress ipAddress) {
+        SpecialPermission.check();
+        return AccessController.doPrivileged((PrivilegedAction<CountryResponse>) () ->
+            cache.putIfAbsent(ipAddress, databasePath.toString(), ip -> {
+                try {
+                    return get().country(ip);
+                } catch (AddressNotFoundException e) {
+                    throw new GeoIpProcessor.AddressNotFoundRuntimeException(e);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }));
+    }
+
+    AsnResponse getAsn(InetAddress ipAddress) {
+        SpecialPermission.check();
+        return AccessController.doPrivileged((PrivilegedAction<AsnResponse>) () ->
+            cache.putIfAbsent(ipAddress, databasePath.toString(), ip -> {
+                try {
+                    return get().asn(ip);
+                } catch (AddressNotFoundException e) {
+                    throw new GeoIpProcessor.AddressNotFoundRuntimeException(e);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }));
+    }
```

Comment:
Maybe extract common functionality to separate method:
```java
private <T extends AbstractResponse> T getResponse(InetAddress ipAddress, CheckedBiFunction<DatabaseReader, InetAddress, T,
    Exception> responseProvider) {
    SpecialPermission.check();
    return AccessController.doPrivileged((PrivilegedAction<T>) () ->
        cache.putIfAbsent(ipAddress, databasePath.toString(), ip -> {
            try {
                return responseProvider.apply(get(), ipAddress);
            } catch (AddressNotFoundException e) {
                throw new GeoIpProcessor.AddressNotFoundRuntimeException(e);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }));
}

CityResponse getCity(InetAddress ipAddress) {
    return getResponse(ipAddress, DatabaseReader::city);
}

CountryResponse getCountry(InetAddress ipAddress) {
    return getResponse(ipAddress, DatabaseReader::country);
}

AsnResponse getAsn(InetAddress ipAddress) {
    return getResponse(ipAddress, DatabaseReader::asn);
}
```

Reply:
Good idea!